redfish-core/lib/environment_metrics.hpp

```diff
@@ -41,6 +167,217 @@ inline void handleEnvironmentMetricsHead(
+        *crow::connections::systemBus, object.begin()->first,
+        "/xyz/openbmc_project/sensors/power/total_power",
+        "xyz.openbmc_project.Sensor.Value", "Value",
+        std::bind_front(afterGetPropertyForPowerWatts, asyncResp, chassisId));
+}
+
+inline void afterGetPowerWatts(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& chassisId, const boost::system::error_code& ec,
+    const dbus::utility::MapperGetSubTreePathsResponse& paths)
+{
+    if (ec)
+    {
+        if (ec.value() != EBADR)
+        {
+            BMCWEB_LOG_ERROR("DBUS response error for getSubTreePaths: {}",
+                             ec.value());
+            messages::internalError(asyncResp->res);
+        }
+        return;
+    }
+
+    auto find = std::ranges::find_if(paths, [](const std::string& tempPath) {
+        sdbusplus::message::object_path path(tempPath);
+        std::string leaf = path.filename();
+        return leaf == "total_power";
+    });
+    if (find == paths.end())
+    {
+        BMCWEB_LOG_DEBUG("There is no total_power");
+        return;
+    }
+
+    dbus::utility::getDbusObject(
+        "/xyz/openbmc_project/sensors/power/total_power", {},
+        std::bind_front(afterGetDbusObjectForPowerWatts, asyncResp, chassisId));
+}
+
+inline void getPowerWatts(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                          const std::string& chassisId)
+{
+    constexpr std::array<std::string_view, 1> interfaces = {
+        "xyz.openbmc_project.Sensor.Value"};
+    dbus::utility::getSubTreePaths(
+        "/xyz/openbmc_project/sensors", 0, interfaces,
+        std::bind_front(afterGetPowerWatts, asyncResp, chassisId));
+}
+
+inline void
+    setPowerSetPoint(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                     uint32_t powerCap)
+{
+    BMCWEB_LOG_DEBUG("Set Power Limit Watts Set Point");
+
+    sdbusplus::asio::setProperty(
+        *crow::connections::systemBus, "xyz.openbmc_project.Settings",
+        "/xyz/openbmc_project/control/host0/power_cap",
+        "xyz.openbmc_project.Control.Power.Cap", "PowerCap", powerCap,
+        [asyncResp](const boost::system::error_code& ec) {
+        if (ec)
+        {
+            if (ec.value() != EBADR)
+            {
+                BMCWEB_LOG_ERROR("Failed to set PowerCap: {}", ec.value());
+                messages::internalError(asyncResp->res);
+            }
+            return;
+        }
+    });
+}
+
+inline void
+    setPowerControlMode(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                        const std::string& controlMode)
+{
+    BMCWEB_LOG_DEBUG("Set Power Limit Watts Control Mode");
+    bool powerCapEnable = false;
+    if (controlMode == "Disabled")
+    {
+        powerCapEnable = false;
+    }
+    else if (controlMode == "Automatic")
+    {
+        powerCapEnable = true;
+    }
+    else
+    {
+        BMCWEB_LOG_DEBUG("Power Control Mode  does not support this mode: {}",
```

Comment:
WARNING is best here since a user error (400)

Reply:
Changed.